### PR TITLE
Replace point constants with static class members

### PR DESCRIPTION
**Pull request by sparr**
#### Summary
Infrastructure "Replace point constants with static class members"

#### Purpose of change

Currently whenever someone needs one of the point constants (zero, min, max, invalid, north, north_east, etc) for a coordinate type that hasn't been needed before, they add it to `coordinate_constants.h`. This file has accumulated over 100 such constants, in a hodge podge with no clear pattern to which constants exist and which constants don't and when it's appropriate to add a new one. This can make the codebase confusing to someone who sees `tripoint_abs_omt_foo` and wonders why `tripoint_rel_ms_foo` doesn't exist. This also leads to scenarios where sometimes the constants are used and sometimes they are created from scratch, e.g. `tripoint_bub_ms( tripoint_zero )`.

#### Describe the solution

I have moved the constants into static members of the `point` and `tripoint` classes, then exposed them in `coord_point_ob`. Now every `tripoint_foo_bar` type has `::zero`, `::min`, `::max`, etc. The relative types have `::north`, `::north_east`, etc. The relative 3d types have `::above` and `::below`.

I also added a helper function to check whether a [tri]point equals the `::invalid` constant, which was the original goal of this effort and led me down the rabbit hole of making `point` and `tripoint` behave more compatibly.

I've attempted to update every usage, including various places that were using `_min` but should have been using `_invalid` (which didn't exist in some cases) and are now using `::invalid`.

I updated the tests that use relative constants with absolute coordinate types, to instead create an absolute coordinate from a relative underlying point.

I described these constants in `POINTS_COORDINATES.md`.

#### Describe alternatives you've considered

I attempted to make the individual members conditional using SFINAE and `std::enable_if_t` but this led down a trail of hurdles in templating and working around compiler bugs.
I tried defining the constants inside the coord class definition which worked in clang and gcc but is rejected by MSVC.

#### Testing

Automated tests pass. I am playing with this change.

#### Additional context

I'm not sure how best to break this up into multiple PRs. I could put the old constants back, change just a few files to the new constants, PR that, then continue until they are all changed?

This PR should be squash merged; the commits are just my development travails.

### CODE REVIEW

**Comment by Procyonae** [comment]
Adding these to POINTS_COORDINATES.md wouldn't hurt

File: src/point.h [hunk 1]
```diff
@@ -348,35 +397,35 @@ struct hash<tripoint> {
...
-        point_west, point_east, point_north, point_south
+        point::west, point::east, point::north, point::south
     }};
 
 inline constexpr std::array<point, 5> five_cardinal_directions{{
-        point_west, point_east, point_north, point_south, point_zero
+        point::west, point::east, point::north, point::south, point::zero
     }};
 
 inline const std::array<tripoint, 8> eight_horizontal_neighbors = { {
```
Comment (CLIDragon) [inline src/point.h:420]: Should this be `constexpr const`?

File: src/point.h [hunk 2]
```diff
@@ -348,35 +397,35 @@ struct hash<tripoint> {
 } // namespace std
 
 inline constexpr std::array<point, 4> four_adjacent_offsets{{
```
Comment (CLIDragon) [inline src/point.h:399]: These should be `inline constexpr const` I believe.

File: src/activity_actor.cpp [hunk 3]
```diff
@@ -6570,8 +6569,8 @@ void chop_tree_activity_actor::finish( player_activity &act, Character &who )
         const point_rel_ms main_dir = pos.xy() - who.pos_bub().xy();
         const int circle_size = 8;
         static constexpr std::array<point, circle_size> circle = {
```
Comment (CLIDragon) [inline src/activity_actor.cpp:6571]: Should this be `const`?

**Review by CLIDragon — COMMENTED** [review]
Given the changes are trivial and easily verified I feel like it should be okay even as a massive line diff.

**Comment by PatrikLundell** [comment]
This looks like a good change.

There was actually a sort of intention behind the set of constants that are defined and the ones that weren't, and that was that only it only made sense to have offsets for the relative types, with zero being used fairly frequently for absolute ones as well. However, it's only mildly detrimental to define everything for all types (detrimental, because it makes it easier to screw up and use an absolute offset constant when you really should have used a relative type, and similar errors).

**Comment by Night-Pryanik** [comment]
@sparr could you please resolve conflicts?

**Comment by sparr** [comment]
@Night-Pryanik I've done a few rounds of conflict resolution on my two big open PRs in order to get tests to pass, but then they sit for a week without review, so I kinda burned out on it. I'd be happy to do it one more time, but I can't keep it up.

**Comment by Night-Pryanik** [comment]
I'm trying my best to review all prs that were sitting for a long time and merge them if possible. I understand it's frustrating to do a big work and got no feedback for it for weeks. In any case I appreciate the work you've done here.

**Comment by PatrikLundell** [comment]
Unfortunately, that's what happens to large PRs. When they get large reviewers shy away from them. It certainly doesn't help that they rarely pass all tests, since there's usually at least one test that fails for unrelated reasons most times for most PRs.
I guess the moral of the story is to cut down PRs to a manageable size or be prepared to wait for weeks or months while fixing merge conflicts as a reviewer finally musters the courage and stamina to tackle the PR.
Note that I'm not blaming you: this is a common situation for me as I've been involved in a number of series of PRs of rather substantial size, and so is very familiar with the situation. As you can guess, I've ended up opting for the large and slow approach much of the time.

**Comment by sparr** [comment]
Updated the summary. Compiling and running tests locally now to check my conflict resolutions.

**Pull request merged**